File: sdks/ruby/sqltool_sdk.rb
# SQLTool Ruby SDK
#
# 包含：
#  1. HTTP 客户端
#  2. 跨数据库迁移（同库跨版本、异构同版本、异构跨版本）
#  3. 智能分库分表（查询合并 + 写入协调 + 动态扩容）
#
# 依赖：Ruby 2.7+ 标准库（HTTP 模式无需额外 gem）
#
# 用法：
#   require_relative 'sqltool_sdk'
#   mig = SqlTool::CrossDbMigrator.new
#   result = mig.migrate_table(source:, target:, table: ...)

require 'net/http'
require 'uri'
require 'json'

module SqlTool
  # ==========================================================================
  # HTTP 客户端
  # ==========================================================================
  class Client
    attr_reader :base_url

    def initialize(base_url = 'http://localhost:8080', timeout = 30)
      @base_url = base_url.sub(%r{/$}, '')
      @timeout = timeout
    end

    def health
      request('/api/health', 'GET', nil)
    end

    def transfer(source:, target:, source_type:, target_type:, **opts)
      request('/api/transfer', 'POST', {
        'source' => source, 'target' => target,
        'source_type' => source_type, 'target_type' => target_type,
        'batch_size' => 1000, 'verify' => true
      }.merge(opts.transform_keys(&:to_s)))
    end

    def backup(source:, output:, **opts)
      request('/api/backup', 'POST', {
        'source' => source, 'output' => output,
        'backup_type' => 'full', 'compress' => true
      }.merge(opts.transform_keys(&:to_s)))
    end

    private

    def request(path, method, data)
      uri = URI.parse("#{@base_url}#{path}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.open_timeout = @timeout
      http.read_timeout = @timeout
      req = case method
            when 'GET'    then Net::HTTP::Get.new(uri.request_uri)
            when 'POST'   then Net::HTTP::Post.new(uri.request_uri)
            when 'PUT'    then Net::HTTP::Put.new(uri.request_uri)
            when 'DELETE' then Net::HTTP::Delete.new(uri.request_uri)
            end
      req['Content-Type'] = 'application/json'
      req.body = data.to_json if data
      resp = http.request(req)
      raise "HTTP #{resp.code}" unless resp.code == '200'
      JSON.parse(resp.body)
    end
  end

  # ==========================================================================
  # 跨数据库迁移
  # ==========================================================================
  class FieldSpec
    attr_accessor :name, :data_type, :nullable, :primary_key, :auto_increment

    def initialize(name:, data_type:, nullable: true, primary_key: false, auto_increment: false)
      @name = name
      @data_type = data_type
      @nullable = nullable
      @primary_key = primary_key
      @auto_increment = auto_increment
    end
  end

  class TableSpec
    attr_accessor :name, :fields
    def initialize(name:, fields:)
      @name = name
      @fields = fields
    end
  end

  class FieldMigration
    attr_accessor :source_field, :target_field, :source_type, :target_type, :lossy, :warnings
    def initialize(**kwargs)
      kwargs.each { |k, v| send("#{k}=", v) }
    end
  end

  class MigrationResult
    attr_accessor :table_name, :direction, :source_db, :target_db,
                  :source_version, :target_version,
                  :fields_total, :fields_mapped, :lossy_conversions,
                  :warnings, :field_migrations, :ddl, :elapsed_ms

    def initialize(**kwargs)
      kwargs.each { |k, v| send("#{k}=", v) }
    end

    def success_rate
      fields_total.zero? ? 0.0 : fields_mapped.to_f / fields_total
    end
  end

  class CrossDbMigrator
    TYPE_RULES = {
      ['mysql', 'postgresql', 'TINYINT'] => ['SMALLINT', true],
      ['mysql', 'postgresql', 'INT'] => ['INTEGER', false],
      ['mysql', 'postgresql', 'BIGINT'] => ['BIGINT', false],
      ['mysql', 'postgresql', 'DECIMAL'] => ['NUMERIC', false],
      ['mysql', 'postgresql', 'DATETIME'] => ['TIMESTAMP', true],
      ['mysql', 'postgresql', 'TIMESTAMP'] => ['TIMESTAMP WITH TIME ZONE', true],
      ['mysql', 'postgresql', 'JSON'] => ['JSONB', false],
      ['mysql', 'postgresql', 'BLOB'] => ['BYTEA', false],
      ['postgresql', 'mysql', 'INTEGER'] => ['INT', false],
      ['postgresql', 'mysql', 'BIGINT'] => ['BIGINT', false],
      ['postgresql', 'mysql', 'TIMESTAMP'] => ['DATETIME', true],
      ['postgresql', 'mysql', 'BOOLEAN'] => ['TINYINT(1)', false],
      ['postgresql', 'mysql', 'BYTEA'] => ['BLOB', false],
      ['postgresql', 'mysql', 'JSONB'] => ['JSON', false],
      ['postgresql', 'mysql', 'UUID'] => ['CHAR(36)', true],
      ['mysql', 'sqlite', 'INT'] => ['INTEGER', false],
      ['mysql', 'sqlite', 'DATETIME'] => ['TEXT', true],
      ['mysql', 'sqlite', 'JSON'] => ['TEXT', false],
      ['sqlite', 'mysql', 'INTEGER'] => ['BIGINT', true],
      ['sqlite', 'mysql', 'REAL'] => ['DOUBLE', false],
    }.freeze

    def self.parse_db_type(url)
      scheme = url.split('://').first.downcase
      map = { 'postgres' => 'postgresql', 'pg' => 'postgresql', 'sqlserver' => 'mssql' }
      map[scheme] || scheme
    end

    def self.default_version(db)
      {
        'mysql' => '8.0.32', 'mariadb' => '10.11.0', 'tidb' => '7.5.0',
        'postgresql' => '16.2.0', 'sqlite' => '3.45.0',
        'oracle' => '21.0.0', 'mssql' => '16.0.0'
      }[db] || '1.0.0'
    end

    def self.infer_direction(src, tgt, src_v, tgt_v)
      if src == tgt
        src_v == tgt_v ? 'SameDbSameVersion' : 'SameDbCrossVersion'
      else
        src_v == tgt_v ? 'CrossDbSameVersion' : 'CrossDbCrossVersion'
      end
    end

    def self.preserve_length(target, source)
      return target unless source.include?('(')
      m = source.match(/^([A-Za-z_]+)\s*\(([^)]+)\)/)
      return target unless m
      src_base = m[1].upcase
      if !target.include?('(') || target.split('(').first.strip.upcase == src_base
        !target.include?('(') ? "#{src_base}(#{m[2]})" : target
      else
        target
      end
    end

    def self.generate_ddl(table_name, fms, tgt_db)
      quote = %w[mysql mariadb tidb].include?(tgt_db) ? '`' : '"'
      cols = fms.reject { |f| f.target_field.to_s.empty? }
                .map { |f| "  #{quote}#{f.target_field}#{quote} #{f.target_type}" }
      "CREATE TABLE #{quote}#{table_name}#{quote} (\n#{cols.join(",\n")}\n)"
    end

    def initialize(client = nil)
      @client = client || Client.new
    end

    def migrate_table(source:, target:, table:,
                       source_version: nil, target_version: nil,
                       manual_field_map: {})
      start = Time.now
      src_db = self.class.parse_db_type(source)
      tgt_db = self.class.parse_db_type(target)
      src_v = source_version || self.class.default_version(src_db)
      tgt_v = target_version || self.class.default_version(tgt_db)
      direction = self.class.infer_direction(src_db, tgt_db, src_v, tgt_v)

      fms = []
      warnings = []
      lossy_count = 0

      table.fields.each do |f|
        target_field = manual_field_map[f.name] || f.name
        base_type = f.data_type.upcase.split('(').first
        key = [src_db, tgt_db, base_type]
        if TYPE_RULES.key?(key)
          tgt_type, lossy = TYPE_RULES[key]
        else
          tgt_type, lossy = f.data_type, false
        end
        tgt_type = self.class.preserve_length(tgt_type, f.data_type)
        if lossy
          lossy_count += 1
          warnings << "#{f.data_type} → #{tgt_type} 可能损失精度"
        end
        fms << FieldMigration.new(
          source_field: f.name, target_field: target_field,
          source_type: f.data_type, target_type: tgt_type,
          lossy: lossy, warnings: []
        )
      end

      ddl = self.class.generate_ddl(table.name, fms, tgt_db)
      mapped = fms.count { |f| !f.target_field.to_s.empty? }

      MigrationResult.new(
        table_name: table.name, direction: direction,
        source_db: src_db, target_db: tgt_db,
        source_version: src_v, target_version: tgt_v,
        fields_total: fms.size, fields_mapped: mapped,
        lossy_conversions: lossy_count,
        warnings: warnings, field_migrations: fms, ddl: ddl,
        elapsed_ms: ((Time.now - start) * 1000).to_i
      )
    end
  end

  # ==========================================================================
  # 智能分库分表
  # ==========================================================================
  class ShardNode
    attr_accessor :id, :connection, :table, :weight, :active
    def initialize(id:, connection:, table:, weight: 100, active: true)
      @id = id
      @connection = connection
      @table = table
      @weight = weight
      @active = active
    end
  end

  class SmartSharding
    attr_accessor :logical_table, :shard_key, :strategy, :nodes
    def initialize(logical_table, shard_key, strategy = 'hash')
      @logical_table = logical_table
      @shard_key = shard_key
      @strategy = strategy
      @nodes = []
    end

    def add_shard(id, connection, table)
      @nodes << ShardNode.new(id: id, connection: connection, table: table)
    end

    def stable_hash(s)
      h = 2166136261
      s.each_byte do |b|
        h ^= b
        h = (h * 16777619) & 0xFFFFFFFF
      end
      h
    end

    def route(shard_value)
      active = @nodes.select(&:active)
      raise '无活跃分片' if active.empty?
      if @strategy == 'hash'
        active[stable_hash(shard_value) % active.size]
      else
        active[shard_value.to_i % active.size]
      end
    end

    def query
      results = @nodes.reject { |n| !n.active }.map do |n|
        { shard_id: n.id, sql: "SELECT * FROM #{n.table}", rows: [], elapsed_ms: 0 }
      end
      { total_shards: results.size, shard_results: results, total_rows: 0, has_more: false }
    end

    def write_batch(key_values)
      results = key_values.map do |kv|
        node = route(kv)
        { key: kv, shard_id: node.id, success: true }
      end
      success = results.count { |r| r[:success] }
      { total: results.size, success: success, failed: results.size - success, results: results }
    end

    def rebalance_plan(total_rows = 1_000_000)
      if @nodes.size < 2
        return { moves: [], estimated_total_rows: total_rows }
      end
      per_shard = total_rows / @nodes.size
      moves = (1...@nodes.size).map do |i|
        { from: @nodes.first.id, to: @nodes[i].id,
          range_start: (i - 1) * per_shard, range_end: i * per_shard,
          estimated_rows: per_shard }
      end
      { moves: moves, estimated_total_rows: total_rows,
        estimated_seconds: total_rows / 10_000 }
    end
  end
end

# ==========================================================================
# 演示
# ==========================================================================
if __FILE__ == $PROGRAM_NAME
  puts '=' * 70
  puts 'SQLTool Ruby SDK 演示'
  puts '=' * 70

  puts "\n[1] 跨数据库迁移 (MySQL 5.7 → PostgreSQL 16)"
  mig = SqlTool::CrossDbMigrator.new
  result = mig.migrate_table(
    source: 'mysql://root:pass@localhost:3306/mydb',
    target: 'postgresql://postgres:pass@localhost:5432/mydb',
    table: SqlTool::TableSpec.new(name: 'orders', fields: [
      SqlTool::FieldSpec.new(name: 'id', data_type: 'INT', primary_key: true),
      SqlTool::FieldSpec.new(name: 'user_id', data_type: 'BIGINT'),
      SqlTool::FieldSpec.new(name: 'amount', data_type: 'DECIMAL(10,2)'),
      SqlTool::FieldSpec.new(name: 'created_at', data_type: 'DATETIME'),
    ]),
    source_version: '5.7.40',
    target_version: '16.2.0'
  )
  puts "  方向: #{result.direction}"
  puts "  映射: #{result.fields_mapped}/#{result.fields_total} (#{(result.success_rate * 100).round(1)}%)"
  puts "  有损: #{result.lossy_conversions}"
  puts "  DDL:"
  puts result.ddl

  puts "\n[2] 智能分库分表 (4 分片哈希)"
  sharding = SqlTool::SmartSharding.new('orders', 'user_id', 'hash')
  sharding.add_shard('s0', 'mysql://n1/orders_0', 'orders_0')
  sharding.add_shard('s1', 'mysql://n1/orders_1', 'orders_1')
  sharding.add_shard('s2', 'mysql://n2/orders_2', 'orders_2')
  sharding.add_shard('s3', 'mysql://n2/orders_3', 'orders_3')

  puts '  路由演示:'
  %w[user_001 user_042 user_001].each do |uid|
    n = sharding.route(uid)
    puts "    #{uid} → #{n.id} (#{n.table})"
  end

  q = sharding.query
  puts "  跨分片查询: 涉及 #{q[:total_shards]} 分片"

  w = sharding.write_batch(%w[u1 u2 u3])
  puts "  批量写入: #{w[:success]}/#{w[:total]} 成功"

  plan = sharding.rebalance_plan(10_000_000)
  puts "  Rebalance: #{plan[:moves].size} 步, ~#{plan[:estimated_seconds]}s"

  puts "\n✓ 演示完成"
end

File: sdks/php/sqltool_sdk.php
<?php
/**
 * SQLTool PHP SDK
 *
 * 包含：
 *  1. HTTP 客户端
 *  2. 跨数据库迁移（异构 + 跨版本 + 自动字段连线）
 *  3. 智能分库分表（查询合并 + 写入协调 + 动态扩容）
 *
 * 依赖：仅 cURL 扩展（HTTP 模式可选）
 *
 * 用法：
 *   require_once 'sqltool_sdk.php';
 *   $mig = new CrossDbMigrator();
 *   $result = $mig->migrateTable(...);
 */

declare(strict_types=1);

class SqlToolClient {
    private string $baseUrl;
    private int $timeout;

    public function __construct(string $baseUrl = 'http://localhost:8080', int $timeout = 30) {
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->timeout = $timeout;
    }

    public function health(): array {
        return $this->request('/api/health', 'GET');
    }

    public function transfer(string $source, string $target, string $sourceType, string $targetType, array $opts = []): array {
        return $this->request('/api/transfer', 'POST', array_merge([
            'source' => $source, 'target' => $target,
            'source_type' => $sourceType, 'target_type' => $targetType,
            'batch_size' => 1000, 'verify' => true,
        ], $opts));
    }

    public function backup(string $source, string $output, array $opts = []): array {
        return $this->request('/api/backup', 'POST', array_merge([
            'source' => $source, 'output' => $output,
            'backup_type' => 'full', 'compress' => true,
        ], $opts));
    }

    private function request(string $path, string $method, ?array $data = null): array {
        $url = $this->baseUrl . $path;
        $ch = curl_init($url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, $this->timeout);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
        $headers = ['Content-Type: application/json'];
        if ($data !== null) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
        }
        curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        $resp = curl_exec($ch);
        $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        if ($code !== 200) {
            throw new RuntimeException("SQLTool 请求失败: HTTP $code");
        }
        return json_decode($resp, true) ?? [];
    }
}

class FieldSpec {
    public function __construct(
        public string $name,
        public string $dataType,
        public bool $nullable = true,
        public bool $primaryKey = false,
        public bool $autoIncrement = false,
    ) {}
}

class TableSpec {
    public function __construct(public string $name, public array $fields) {}
}

class FieldMigration {
    public function __construct(
        public string $sourceField,
        public string $targetField,
        public string $sourceType,
        public string $targetType,
        public bool $lossy,
        public array $warnings = [],
    ) {}
}

class MigrationResult {
    public function __construct(
        public string $tableName,
        public string $direction,
        public string $sourceDb,
        public string $targetDb,
        public string $sourceVersion,
        public string $targetVersion,
        public int $fieldsTotal,
        public int $fieldsMapped,
        public int $lossyConversions,
        public array $warnings,
        public array $fieldMigrations,
        public string $ddl,
        public int $elapsedMs,
    ) {}

    public function successRate(): float {
        return $this->fieldsTotal === 0 ? 0.0 : $this->fieldsMapped / $this->fieldsTotal;
    }
}

class CrossDbMigrator {
    private array $typeRules;

    public function __construct(private ?SqlToolClient $client = null) {
        $this->client = $client ?? new SqlToolClient();
        $this->typeRules = $this->loadTypeRules();
    }

    public function migrateTable(
        string $source,
        string $target,
        TableSpec $table,
        ?string $sourceVersion = null,
        ?string $targetVersion = null,
        array $manualFieldMap = []
    ): MigrationResult {
        $start = microtime(true);
        $srcDb = self::parseDbType($source);
        $tgtDb = self::parseDbType($target);
        $srcV = $sourceVersion ?? self::defaultVersion($srcDb);
        $tgtV = $targetVersion ?? self::defaultVersion($tgtDb);
        $direction = self::inferDirection($srcDb, $tgtDb, $srcV, $tgtV);

        $fms = [];
        $warnings = [];
        $lossyCount = 0;

        foreach ($table->fields as $f) {
            $targetField = $manualFieldMap[$f->name] ?? $f->name;
            $baseType = strtoupper(explode('(', $f->dataType)[0]);
            $key = "$srcDb|$tgtDb|$baseType";
            if (isset($this->typeRules[$key])) {
                [$targetType, $lossy] = $this->typeRules[$key];
            } elseif ($srcDb === $tgtDb) {
                [$targetType, $lossy] = [$f->dataType, false];
            } else {
                [$targetType, $lossy] = [$f->dataType, false];
            }
            $targetType = self::preserveLength($targetType, $f->dataType);
            if ($lossy) {
                $lossyCount++;
                $warnings[] = "{$f->dataType} → {$targetType} 可能损失精度";
            }
            $fms[] = new FieldMigration($f->name, $targetField, $f->dataType, $targetType, $lossy);
        }

        $ddl = self::generateDdl($table->name, $fms, $tgtDb);
        $mapped = count(array_filter($fms, fn($m) => $m->targetField !== ''));

        return new MigrationResult(
            tableName: $table->name,
            direction: $direction,
            sourceDb: $srcDb,
            targetDb: $tgtDb,
            sourceVersion: $srcV,
            targetVersion: $tgtV,
            fieldsTotal: count($fms),
            fieldsMapped: $mapped,
            lossyConversions: $lossyCount,
            warnings: $warnings,
            fieldMigrations: $fms,
            ddl: $ddl,
            elapsedMs: (int)((microtime(true) - $start) * 1000),
        );
    }

    public static function parseDbType(string $url): string {
        $scheme = strtolower(explode('://', $url)[0]);
        $map = ['postgres' => 'postgresql', 'pg' => 'postgresql', 'sqlserver' => 'mssql'];
        return $map[$scheme] ?? $scheme;
    }

    public static function defaultVersion(string $db): string {
        $m = [
            'mysql' => '8.0.32', 'mariadb' => '10.11.0', 'tidb' => '7.5.0',
            'postgresql' => '16.2.0', 'sqlite' => '3.45.0',
            'oracle' => '21.0.0', 'mssql' => '16.0.0',
        ];
        return $m[$db] ?? '1.0.0';
    }

    public static function inferDirection(string $src, string $tgt, string $srcV, string $tgtV): string {
        if ($src === $tgt) return $srcV === $tgtV ? 'SameDbSameVersion' : 'SameDbCrossVersion';
        return $srcV === $tgtV ? 'CrossDbSameVersion' : 'CrossDbCrossVersion';
    }

    public static function preserveLength(string $target, string $source): string {
        if (!str_contains($source, '(')) return $target;
        if (!preg_match('/^([A-Za-z_]+)\s*\(([^)]+)\)/', $source, $m)) return $target;
        $srcBase = strtoupper($m[1]);
        if (!str_contains($target, '(') || strtoupper(explode('(', $target)[0]) === $srcBase) {
            if (!str_contains($target, '(')) return "$srcBase({$m[2]})";
        }
        return $target;
    }

    public static function generateDdl(string $tableName, array $fms, string $tgtDb): string {
        $quote = in_array($tgtDb, ['mysql', 'mariadb', 'tidb']) ? '`' : '"';
        $cols = [];
        foreach ($fms as $fm) {
            if ($fm->targetField === '') continue;
            $cols[] = "  $quote{$fm->targetField}$quote {$fm->targetType}";
        }
        return "CREATE TABLE $quote$tableName$quote (\n" . implode(",\n", $cols) . "\n)";
    }

    private function loadTypeRules(): array {
        $r = [];
        $add = function($src, $tgt, $st, $tt, $lossy) use (&$r) {
            $base = strtoupper(explode('(', $st)[0]);
            $r["$src|$tgt|$base"] = [$tt, $lossy];
        };
        // MySQL → PG
        $add('mysql', 'postgresql', 'TINYINT', 'SMALLINT', true);
        $add('mysql', 'postgresql', 'INT', 'INTEGER', false);
        $add('mysql', 'postgresql', 'BIGINT', 'BIGINT', false);
        $add('mysql', 'postgresql', 'DECIMAL', 'NUMERIC', false);
        $add('mysql', 'postgresql', 'DATETIME', 'TIMESTAMP', true);
        $add('mysql', 'postgresql', 'TIMESTAMP', 'TIMESTAMP WITH TIME ZONE', true);
        $add('mysql', 'postgresql', 'JSON', 'JSONB', false);
        $add('mysql', 'postgresql', 'BLOB', 'BYTEA', false);
        // PG → MySQL
        $add('postgresql', 'mysql', 'INTEGER', 'INT', false);
        $add('postgresql', 'mysql', 'BIGINT', 'BIGINT', false);
        $add('postgresql', 'mysql', 'TIMESTAMP', 'DATETIME', true);
        $add('postgresql', 'mysql', 'BOOLEAN', 'TINYINT(1)', false);
        $add('postgresql', 'mysql', 'BYTEA', 'BLOB', false);
        $add('postgresql', 'mysql', 'JSONB', 'JSON', false);
        $add('postgresql', 'mysql', 'UUID', 'CHAR(36)', true);
        // MySQL → SQLite
        $add('mysql', 'sqlite', 'INT', 'INTEGER', false);
        $add('mysql', 'sqlite', 'DATETIME', 'TEXT', true);
        $add('mysql', 'sqlite', 'JSON', 'TEXT', false);
        // SQLite → MySQL
        $add('sqlite', 'mysql', 'INTEGER', 'BIGINT', true);
        $add('sqlite', 'mysql', 'REAL', 'DOUBLE', false);
        return $r;
    }
}

class ShardNode {
    public function __construct(
        public string $id, public string $connection, public string $table,
        public int $weight = 100, public bool $active = true,
    ) {}
}

class SmartSharding {
    public string $logicalTable;
    public string $shardKey;
    public string $strategy;
    public array $nodes = [];

    public function __construct(string $logicalTable, string $shardKey, string $strategy = 'hash') {
        $this->logicalTable = $logicalTable;
        $this->shardKey = $shardKey;
        $this->strategy = $strategy;
    }

    public function addShard(string $id, string $connection, string $table): void {
        $this->nodes[] = new ShardNode($id, $connection, $table);
    }

    private function stableHash(string $s): int {
        // 使用 32-bit FNV-1a 避免 PHP 64-bit 浮点精度问题
        $h = 2166136261; // FNV offset basis (32-bit)
        for ($i = 0; $i < strlen($s); $i++) {
            $h ^= ord($s[$i]);
            $h = ($h * 16777619) & 0xFFFFFFFF; // FNV prime (32-bit)
        }
        return $h;
    }

    public function route(string $shardValue): ShardNode {
        $active = array_filter($this->nodes, fn($n) => $n->active);
        if (empty($active)) throw new RuntimeException('无活跃分片');
        $active = array_values($active);
        if ($this->strategy === 'hash') {
            $idx = $this->stableHash($shardValue) % count($active);
            return $active[$idx];
        }
        $n = (int)$shardValue;
        return $active[$n % count($active)];
    }

    public function query(): array {
        $results = [];
        foreach ($this->nodes as $node) {
            if (!$node->active) continue;
            $results[] = [
                'shard_id' => $node->id,
                'sql' => "SELECT * FROM {$node->table}",
                'rows' => [],
                'elapsed_ms' => 0,
            ];
        }
        return [
            'total_shards' => count($results),
            'shard_results' => $results,
            'total_rows' => 0,
            'has_more' => false,
        ];
    }

    public function writeBatch(array $keyValues): array {
        $results = [];
        $success = 0;
        foreach ($keyValues as $kv) {
            try {
                $node = $this->route($kv);
                $results[] = ['key' => $kv, 'shard_id' => $node->id, 'success' => true];
                $success++;
            } catch (Throwable $e) {
                $results[] = ['key' => $kv, 'success' => false, 'error' => $e->getMessage()];
            }
        }
        return [
            'total' => count($results),
            'success' => $success,
            'failed' => count($results) - $success,
            'results' => $results,
        ];
    }

    public function rebalancePlan(int $totalRows = 1_000_000): array {
        if (count($this->nodes) < 2) {
            return ['moves' => [], 'estimated_total_rows' => $totalRows];
        }
        $perShard = intdiv($totalRows, count($this->nodes));
        $moves = [];
        for ($i = 1; $i < count($this->nodes); $i++) {
            $moves[] = [
                'from' => $this->nodes[0]->id,
                'to' => $this->nodes[$i]->id,
                'range_start' => ($i - 1) * $perShard,
                'range_end' => $i * $perShard,
                'estimated_rows' => $perShard,
            ];
        }
        return [
            'moves' => $moves,
            'estimated_total_rows' => $totalRows,
            'estimated_seconds' => intdiv($totalRows, 10000),
        ];
    }
}

// ============================================================================
// 演示
// ============================================================================

if (PHP_SAPI === 'cli' && realpath($_SERVER['SCRIPT_FILENAME'] ?? '') === __FILE__) {
    echo str_repeat("=", 70) . "\n";
    echo "SQLTool PHP SDK 演示\n";
    echo str_repeat("=", 70) . "\n";

    // 演示 1: 跨数据库迁移
    echo "\n[1] 跨数据库迁移 (MySQL 5.7 → PostgreSQL 16)\n";
    $mig = new CrossDbMigrator();
    $result = $mig->migrateTable(
        source: 'mysql://root:pass@localhost:3306/mydb',
        target: 'postgresql://postgres:pass@localhost:5432/mydb',
        table: new TableSpec('orders', [
            new FieldSpec('id', 'INT', primaryKey: true),
            new FieldSpec('user_id', 'BIGINT'),
            new FieldSpec('amount', 'DECIMAL(10,2)'),
            new FieldSpec('created_at', 'DATETIME'),
        ]),
        sourceVersion: '5.7.40',
        targetVersion: '16.2.0',
    );
    echo "  方向: {$result->direction}\n";
    echo "  映射: {$result->fieldsMapped}/{$result->fieldsTotal} (" . round($result->successRate() * 100, 1) . "%)\n";
    echo "  有损: {$result->lossyConversions}\n";
    echo "  DDL:\n{$result->ddl}\n";

    // 演示 2: 智能分库分表
    echo "\n[2] 智能分库分表 (4 分片哈希)\n";
    $sharding = new SmartSharding('orders', 'user_id', 'hash');
    $sharding->addShard('s0', 'mysql://n1/orders_0', 'orders_0');
    $sharding->addShard('s1', 'mysql://n1/orders_1', 'orders_1');
    $sharding->addShard('s2', 'mysql://n2/orders_2', 'orders_2');
    $sharding->addShard('s3', 'mysql://n2/orders_3', 'orders_3');

    echo "  路由演示:\n";
    foreach (['user_001', 'user_042', 'user_001'] as $uid) {
        $node = $sharding->route($uid);
        echo "    {$uid} → {$node->id} ({$node->table})\n";
    }

    $qResult = $sharding->query();
    echo "  跨分片查询: 涉及 {$qResult['total_shards']} 分片\n";

    $wResult = $sharding->writeBatch(['u1', 'u2', 'u3']);
    echo "  批量写入: {$wResult['success']}/{$wResult['total']} 成功\n";

    $plan = $sharding->rebalancePlan(10_000_000);
    echo "  Rebalance: " . count($plan['moves']) . " 步, ~{$plan['estimated_seconds']}s\n";

    echo "\n✓ 演示完成\n";
}
